core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java

```diff
@@ -0,0 +1,393 @@
+        private final AtomicInteger totalSentOps = new AtomicInteger();
+        private final AtomicInteger totalSkippedOps = new AtomicInteger();
+        private AtomicBoolean closed = new AtomicBoolean();
+        private AtomicBoolean runOnce = new AtomicBoolean();
+
+        SnapshotSender(Logger logger, SyncAction syncAction, ResyncTask task, ShardId shardId, String primaryAllocationId,
+                       Translog.Snapshot snapshot, int chunkSizeInBytes, long startingSeqNo, long endingSeqNo,
+                       ActionListener<Void> listener) {
+            this.logger = logger;
+            this.syncAction = syncAction;
+            this.task = task;
+            this.shardId = shardId;
+            this.primaryAllocationId = primaryAllocationId;
+            this.snapshot = snapshot;
+            this.chunkSizeInBytes = chunkSizeInBytes;
+            this.startingSeqNo = startingSeqNo;
+            this.endingSeqNo = endingSeqNo;
+            this.listener = listener;
+            task.setTotalOperations(snapshot.totalOperations());
+        }
+
+        @Override
+        public void onResponse(ResyncReplicationResponse response) {
+            run();
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (closed.compareAndSet(false, true)) {
+                listener.onFailure(e);
+            }
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            long size = 0;
+            final List<Translog.Operation> operations = new ArrayList<>();
+
+            task.setPhase("syncing");
+            task.setResyncedOperations(totalSentOps.get());
+            task.setSkippedOperations(totalSkippedOps.get());
+
+            Translog.Operation operation;
+            while ((operation = snapshot.next()) != null) {
+                final long seqNo = operation.seqNo();
+                if (startingSeqNo >= 0 &&
+                    (seqNo == SequenceNumbersService.UNASSIGNED_SEQ_NO || seqNo < startingSeqNo || seqNo > endingSeqNo)) {
+                    totalSkippedOps.incrementAndGet();
+                    continue;
+                }
+                operations.add(operation);
+                size += operation.estimateSize();
+                totalSentOps.incrementAndGet();
+
+                // check if this request is past bytes threshold, and if so, send it off
+                if (size >= chunkSizeInBytes) {
+                    break;
+                }
+            }
+
+            if (runOnce.compareAndSet(false, true) || !operations.isEmpty()) {
+                ResyncReplicationRequest request = new ResyncReplicationRequest(shardId, operations);
+                logger.trace("sending batch of [{}][{}] (total sent: [{}], skipped: [{}])", operations.size(), new ByteSizeValue(size),
+                    totalSentOps.get(), totalSkippedOps.get());
+                syncAction.sync(request, task, primaryAllocationId, this);
```

Comment:
that would just indicate that the first ops have gone out, which can also be tracked with the detailed tasks stats. I don't see the benefit.